p2p/src/peer.rs

```diff
@@ -38,39 +38,125 @@ const MAX_HANDSHAKE_LENGTH: usize = 255;
+    secret_key: PrivateKey,
     /// Public part of keypair
-    pub public_key: PublicKey,
+    public_key: PublicKey,
     /// Encryptor created from session key, that we got by Diffie-Hellman scheme
-    pub cipher: Option<SymmetricEncryptor<E>>,
-    /// The to send received messages upstairs
-    pub broker: Broker,
-    /// A flag that stops listening stream
-    finish_sender: Option<Sender<()>>,
+    cipher: Option<SymmetricEncryptor<E>>,
     /// Phantom
-    pub _key_exchange: PhantomData<K>,
+    _key_exchange: PhantomData<K>,
     /// Phantom2
-    pub _post_type: PhantomData<T>,
+    _post_type: PhantomData<T>,
+}
+
+impl<T, K, E> Cryptographer<T, K, E>
+where
+    T: Debug + Encode + Decode + Send + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    fn default_or_err() -> Result<Self, Error> {
+        // P2P encryption primitives
+        let dh = K::new();
+        let (public_key, secret_key) = dh.keypair(None).log_warn("Error generating keypair")?;
+        Ok(Self {
+            secret_key,
+            public_key,
+            cipher: None,
+            _key_exchange: PhantomData::default(),
+            _post_type: PhantomData::default(),
+        })
+    }
+
+    fn decrypt(&self, data: Vec<u8>) -> Result<Vec<u8>, Error> {
+        match &self.cipher {
+            None => Ok(data),
+            Some(cipher) => Ok(cipher.decrypt_easy(DEFAULT_AAD.as_ref(), data.as_slice())?),
+        }
+    }
+
+    fn encrypt(&self, data: Vec<u8>) -> Result<Vec<u8>, Error> {
+        match &self.cipher {
+            None => Ok(data),
+            Some(cipher) => Ok(cipher.encrypt_easy(DEFAULT_AAD.as_ref(), data.as_slice())?),
+        }
+    }
+
+    fn new_encryptor(key: &[u8]) -> Result<SymmetricEncryptor<E>, aead::Error> {
+        SymmetricEncryptor::<E>::new_with_key(key)
+    }
+
+    /// Creates a shared key from two public keys - local and external,
+    /// then instantiates an encryptor from that key.
+    fn derive_shared_key(&mut self, public_key: &PublicKey) -> Result<&Self, Error> {
+        let dh = K::new();
+        let shared = dh
+            .compute_shared_secret(&self.secret_key, public_key)
+            .log_warn("Error creating shared secret")?;
+        debug!("Derived shared key: {:?}", &shared.0);
+        let encryptor = Cryptographer::<T, K, E>::new_encryptor(shared.0.as_slice())
+            .log_warn("Error creating encryptor")?;
+        self.cipher = Some(encryptor);
+        Ok(self)
+    }
+}
+
+/// This is an endpoint, that juggles messages between [`crate::Network`] and another connected node.
+pub struct Peer<T, K, E>
+where
+    T: Debug + Encode + Decode + Send + Clone + 'static,
+    K: KeyExchangeScheme + Send + 'static,
+    E: Encryptor + Send + 'static,
+{
+    /// Peer identification
+    id: PeerId,
+    /// Peer connnection
+    connection: Connection,
+    /// Current peer/connection state
+    state: State,
```

Comment:
I thought "type state" meant state transition as a struct conversion, which replaces enum state.
Maybe there are some barriers to implement like that?

Reply:
Yes. I've detailed them in the Alternative implementations section.

Reply:
I don't get it. Can't we do like this?

```rust
use std::sync::Arc;

#[tokio::main]
async fn main() {
    let _ = S1Peer::default().convert_or_revert().await;
}

#[derive(Default)]
struct NoClone;

#[derive(Clone, Default)]
struct S1Peer {
    no_clone: Arc<NoClone>,
}
struct S2Peer {
    no_clone: Arc<NoClone>,
}

struct Error;

impl S1Peer {
    async fn convert_or_revert(self) -> Result<S2Peer, S1Peer> {
        self.clone().try_convert().await.map_err(|_| self)
    }
    async fn try_convert(mut self) -> Result<S2Peer, Error> {
        self.mutate().await;
        if true {
            return Err(Error);
        }
        Ok(S2Peer {
            no_clone: self.no_clone,
        })
    }
    async fn mutate(&mut self) {}
}
```

Reply:
Consider the amount of boilerplate and the actual benefits. Here you've added two extra structs (not enums) that contain an `Arc` even though technically there should never be more than one instance of `Peer`. 

Right now the `Peer` type is public, but it can only be instantiated via either `new_to` or `new_from` outside the crate. None of the other operations depend on the state, and are not exposed. 

Because of the way that state is propagated, the only difference between querying `self.state` and `match self`, is that in the latter case all internal functions that take a `&mut self` need to use `std::mem::replace`. Can it be done? Absolutely. But IMO it'd be a step backwards.

Reply:
Regarding `&mut self`.
Have you considered externalizing read/write part?
For example:

```rust
use tokio::sync::oneshot;

struct ConnectingPeer {
    //snip
}
struct IncomingPeer {
    // snip
    communicator: oneshot::Sender<StreamMsg>,
}
struct ReadyPeer {
    // snip
    communicator: oneshot::Sender<StreamMsg>,
    crypter: oneshot::Sender<CryptMsg>,
}

struct Msg<T1, T2> {
    body: T1,
    return_addr: oneshot::Sender<T2>,
}

type StreamMsg = Msg<Data, Data>;
struct Communicator {
    // snip
    receiver: oneshot::Receiver<StreamMsg>,
    stream: TcpStream,
}
impl ConnectingPeer {
    async fn connect(self, stream: TcpStream) -> Result<(IncomingPeer, Communicator), Error> {
        // TODO hello
        let (sender, receiver) = oneshot::channel();
        Ok((
            IncomingPeer {
                communicator: sender,
            },
            Communicator { receiver, stream },
        ))
    }
}

type CryptMsg = Msg<(KeyPair, Data), Data>;
struct Crypter {
    // snip
    receiver: oneshot::Receiver<CryptMsg>,
}
impl IncomingPeer {
    async fn handshake(self, keys: KeyPair) -> Result<(ReadyPeer, Crypter), Error> {
        // TODO handshake
        let (sender, receiver) = oneshot::channel();
        Ok((
            ReadyPeer {
                communicator: self.communicator,
                crypter: sender,
            },
            Crypter { receiver },
        ))
    }
}

#[derive(Debug)]
struct Error;

struct Data;

struct KeyPair;
struct TcpStream;

impl Crypter {
    fn spawn(self) {
        // snip
    }
}
impl Communicator {
    fn spawn(self) {
        // snip
    }
}
impl ReadyPeer {
    fn spawn(self) {
        // snip
    }
}

#[tokio::test]
async fn test() {
    let connecting_peer = ConnectingPeer {};
    let stream = TcpStream;
    let (incoming_peer, communicator) = connecting_peer.connect(stream).await.unwrap();
    communicator.spawn();
    let keys = KeyPair;
    let (ready_peer, crypter) = incoming_peer.handshake(keys).await.unwrap();
    crypter.spawn();
    ready_peer.spawn();
}

```